core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,378 @@
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MTX;
+	tmp = (len == 1) ? tmp | I2CR_TX_NO_AK : tmp & ~I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+	i2c_io_read8(bid, I2DR);
+
+	/* dummy read */
+	ret = i2c_read_byte(bid, &dummy);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/*
+	 * i.MX 8M Mini APRM Rev2 08/2019:
+	 * a data transfer ends when the master signals a stop; for a master
+	 * receiver to terminate a transfer it must inform the slave transmiter
+	 * by not acknowledging the last data byte. This is done by setting the
+	 * transmit acknowledge bit before reading the next-to-last byte.
+	 */
+	do {
+		if (len == 2) {
+			tmp = i2c_io_read8(bid, I2CR) | I2CR_TX_NO_AK;
+			i2c_io_write8(bid, I2CR, tmp);
+		}
+
+		ret = i2c_read_byte(bid, buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	} while (len--);
+
+	return ret;
+}
+
+/* i.MX 8M Mini APRM Rev2 08/2019: steps previous to a read/write operation  */
+static TEE_Result i2c_init_transfer(uint8_t bid, uint8_t chip)
+{
+	uint32_t tmp = 0;
+	int ret = 0;
+
+	/* idle the bus */
+	i2c_idle_bus(bid);
+
+	/* enable the interface */
+	i2c_io_write8(bid, I2CR, I2CR_IEN);
+
+	/* 16.1.4.1 start master transaction */
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MSTA;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	/* 16.1.4.2. on multimaster bus system, wait unil we have the bus */
+	ret = i2c_sync_bus(bid, &bus_is_busy, NULL);
+	if (ret < 0)
+		return ret;
+
+	/* put the slave address on the bus */
+	return i2c_write_data(bid, &chip, 1);
+}
+
+TEE_Result imx_i2c_read(uint8_t bid, uint8_t chip, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	ret = i2c_init_transfer(bid, chip << 1 | 1);
+	if (ret == TEE_SUCCESS)
+		ret = i2c_read_data(bid, buf, len);
+	i2c_idle_bus(bid);
+
+	return ret;
+}
+
+TEE_Result imx_i2c_write(uint8_t bid, uint8_t chip, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	ret = i2c_init_transfer(bid, chip << 1);
+	if (ret == TEE_SUCCESS)
+		ret = i2c_write_data(bid, buf, len);
+	i2c_idle_bus(bid);
+
+	return ret;
+}
+
+TEE_Result imx_i2c_probe(uint8_t bid, uint8_t chip)
+{
+	return imx_i2c_write(bid, chip, NULL, 0);
+}
+
+TEE_Result imx_i2c_init(uint8_t bid, int rate)
+{
+	struct imx_i2c_mux *mux = &i2c_mux;
+
+	if (bid > ARRAY_SIZE(i2c_bus) - 1)
+		return TEE_ERROR_GENERIC;
+
+	if (!rate)
+		return TEE_ERROR_GENERIC;
+
+	/* 16.1.4 Configure iomuxc */
+	io_write32(mux->base.va + mux->i2c[bid].scl_mux, I2C_MUX_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].scl_cfg, I2C_CFG_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].sda_mux, I2C_MUX_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].sda_cfg, I2C_CFG_VAL(bid));
+
+	/* 16.1.4.1 set the data sampling rate */
+	i2c_set_bus_speed(bid, rate);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_va(paddr_t pa, vaddr_t *va)
+{
+	*va = (vaddr_t)phys_to_virt(pa, MEM_AREA_IO_SEC);
+	if (*va)
+		return TEE_SUCCESS;
+
+	if (!core_mmu_add_mapping(MEM_AREA_IO_SEC, pa, 0xFFFF))
```

Comment:
I suggest `0x10000` instead of `0xFFFF`. Argument is a byte size, not an inclusive end offset.

By the way, maybe you meant 0xFFF for a 4KB page, not 0xFFFF (16KB)?

Reply:
just grabbed the space from the kernel's imx8mm.dtsi  (I plan to add DT support soon, is just trivial at least to avoid the hard-coding of i2c base address on those systems using it)